src/indexable.cr

```diff
@@ -170,6 +170,311 @@ module Indexable(T)
     (0...size).bsearch { |index| yield unsafe_fetch(index), index }
   end
 
+  # Returns an `Array` of all ordered combinations of elements taken from each
+  # of `self` and *others* as `Tuple`s.
+  # Traversal of elements starts from the last `Indexable` argument.
+  #
+  # ```
+  # [1, 2, 3].cartesian_product({'a', 'b'})                     # => [{1, 'a'}, {1, 'b'}, {2, 'a'}, {2, 'b'}, {3, 'a'}, {3, 'b'}]
+  # ['a', 'b'].cartesian_product({1, 2}, {'c', 'd'}).map &.join # => ["a1c", "a1d", "a2c", "a2d", "b1c", "b1d", "b2c", "b2d"]
+  # ```
+  def cartesian_product(*others : Indexable)
+    capacity = others.product(size, &.size)
+    result = Array(typeof(cartesian_product_type(*others))).new(capacity)
+    each_cartesian(*others) do |product|
+      result << product
+    end
+    result
+  end
+
+  private def cartesian_product_type(*others)
+    v = each_cartesian(*others).next
+    raise "" if v.is_a?(Iterator::Stop)
+    v
+  end
+
+  # Returns an `Array` of all ordered combinations of elements taken from each
+  # of the *indexables* as `Array`s.
+  # Traversal of elements starts from the last `Indexable`.
+  #
+  # `#cartesian_product` is preferred over this class method when the quantity
+  # of *indexables* is known in advance.
+  #
+  # ```
+  # Indexable.cartesian_product([[1, 2, 3], [4, 5]]) # => [[1, 4], [1, 5], [2, 4], [2, 5], [3, 4], [3, 5]]
+  # ```
+  def self.cartesian_product(indexables : Indexable(Indexable))
+    capacity = indexables.product(&.size)
```

Comment:
Do you need to give a more readable error in the empty case?

Reply:
Actually the current code is wrong; if *indexables* is empty then it should return `[[]]` of an appropriate type, it's only when *indexables* is non-empty and contains an empty collection that the method returns `[]`. Passing empty collections to `cartesian_product` should not be an error.